File: Python/OrquestratedByClient.py
# State
class IDrivingMode:
    def throttle(self):
        pass

    def brake(self):
        pass

# Context
class Car:
    def __init__(self):
        self.speed = 0
        self._driving_mode = EcoDrivingMode(self)

    def throttle(self):
        self._driving_mode.throttle()

    def brake(self):
        self._driving_mode.brake()

    def driving_mode(self, mode):
        self._driving_mode = mode

# Concrete States
class EcoDrivingMode(IDrivingMode):
    def __init__(self, car):
        self._car = car

    def throttle(self):
        self._car.speed += 5
        print(f"EcoDrivingMode -> throttle: {self._car.speed}km/h")

    def brake(self):
        self._car.speed -= 5
        self._car.speed = max(self._car.speed, 0)
        print(f"EcoDrivingMode -> brake: {self._car.speed}km/h")

class ComfortDrivingMode(IDrivingMode):
    def __init__(self, car):
        self._car = car

    def throttle(self):
        self._car.speed += 10
        print(f"ComfortDrivingMode -> throttle: {self._car.speed}km/h")

    def brake(self):
        self._car.speed -= 10
        self._car.speed = max(self._car.speed, 0)
        print(f"ComfortDrivingMode -> brake: {self._car.speed}km/h")

class SportDrivingMode(IDrivingMode):
    def __init__(self, car):
        self._car = car

    def throttle(self):
        self._car.speed += 15
        print(f"SportDrivingMode -> throttle: {self._car.speed}km/h")

    def brake(self):
        self._car.speed -= 15
        self._car.speed = max(self._car.speed, 0)
        print(f"SportDrivingMode -> brake: {self._car.speed}km/h")

# Client
car = Car()

car.brake()
car.throttle()
car.throttle()
car.brake()
car.throttle()
car.throttle()

car.driving_mode(ComfortDrivingMode(car))

car.brake()
car.throttle()
car.throttle()
car.throttle()
car.brake()

car.driving_mode(SportDrivingMode(car))
car.brake()
car.throttle()
car.throttle()
car.throttle()
car.brake()

File: TypeScript/OrquestratedByClient.ts
// State
interface IDrivingMode
{
    throttle(): void;
    brake(): void;
}

// Context
class Car
{
    public speed: number;
    private _drivingMode: IDrivingMode;

    constructor()
    {
        this.speed = 0;
        this._drivingMode = new EcoDrivingMode(this);
    }

    public throttle(): void
    {
        this._drivingMode.throttle();
    }

    public brake(): void
    {
        this._drivingMode.brake();
    }

    public set drivingMode(mode: IDrivingMode)
    {
        this._drivingMode = mode;
    }
}

// Concrete States
class EcoDrivingMode implements IDrivingMode
{
    constructor(private car: Car) {}

    throttle(): void
    {
        this.car.speed += 5;
        console.log(`EcoDrivingMode -> throttle: ${this.car.speed}km/h`);
    }

    brake(): void
    {
        this.car.speed -= 5;
        if (this.car.speed < 0) this.car.speed = 0;
        console.log(`EcoDrivingMode -> brake: ${this.car.speed}km/h`);
    }
}

// Concrete States
class ComfortDrivingMode implements IDrivingMode
{
    constructor(private car: Car) {}

    throttle(): void
    {
        this.car.speed += 10;
        console.log(`ComfortDrivingMode -> throttle: ${this.car.speed}km/h`);
    }

    brake(): void
    {
        this.car.speed -= 10;
        if (this.car.speed < 0) this.car.speed = 0;
        console.log(`ComfortDrivingMode -> brake: ${this.car.speed}km/h`);
    }
}

// Concrete States
class SportDrivingMode implements IDrivingMode
{
    constructor(private car: Car) {}

    throttle(): void
    {
        this.car.speed += 15;
        console.log(`SportDrivingMode -> throttle: ${this.car.speed}km/h`);
    }

    brake(): void
    {
        this.car.speed -= 15;
        if (this.car.speed < 0) this.car.speed = 0;
        console.log(`SportDrivingMode -> brake: ${this.car.speed}km/h`);
    }
}

// Client
var car = new Car();

car.brake();
car.throttle();
car.throttle();
car.brake();
car.throttle();
car.throttle();

car.drivingMode = new ComfortDrivingMode(car);

car.brake();
car.throttle();
car.throttle();
car.throttle();
car.brake();

car.drivingMode = new SportDrivingMode(car);
car.brake();
car.throttle();
car.throttle();
car.throttle();
car.brake();
